### PR TITLE
[SMTChecker] Added type constraints on transaction variables also for the contract deployment phase.

### LINKED ISSUE #10703: [SMTChecker] Missing type constraints on transaction variables in contract deployment phase.
The following examples current produce false reports from SMTChecker. The reason are missing type constraints on transaction variables in contract deployment phase.

```
pragma experimental SMTChecker;

contract C {
	uint x;

	constructor() {
		x = block.timestamp + 0;
	}
}
```
produces
```
Warning: CHC: Overflow (resulting value larger than 2**256 - 1) happens here.
Counterexample:
x = 0


Transaction trace:
constructor()
 --> tx2.sol:7:7:
  |
7 | 		x = block.timestamp + 0;
  | 		    ^^^^^^^^^^^^^^^^^^^

```
even though the overflow is not possible.

```
pragma experimental SMTChecker;

contract C {
	address owner;

	constructor() {
		owner = msg.sender;
        	assert(owner >= address(0));
	}
}
```
produces 
```
Warning: CHC: Assertion violation happens here.
Counterexample:
owner = (- 1)



Transaction trace:
constructor()
 --> tx.sol:8:9:
  |
8 |         assert(owner >= address(0));
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

```
even though the assertion is safe and the CEX `owner = (- 1)` does not make sense.

### PR DESCRIPTION
**Pull request by blishko**
This PR adds missing type constraints on transaction variables in the contract deployment phase.

Fixes #10703.

### CODE REVIEW

**Comment by blishko** [comment]
Unfortunately, with this change in the encoding, a bug in Z3 causes segfault on many examples in our tests.
This bug has been reported to Z3 developers some time ago (https://github.com/Z3Prover/z3/issues/4772) and has been fixed already in Z3 master. But we need to wait until a new release of Z3 is out.

File: libsolidity/formal/SymbolicState.h [hunk 1]
```diff
@@ -126,7 +126,7 @@ class SymbolicState
 	smtutil::SortPointer const& txSort() const { return m_tx.sort(); }
 	void newTx() { m_tx.newVar(); }
 	smtutil::Expression txMember(std::string const& _member) const;
-	smtutil::Expression txConstraints(FunctionDefinition const& _function) const;
+	smtutil::Expression txConstraints(FunctionDefinition const* _function) const;
```
Comment (leonardoalt) [inline libsolidity/formal/SymbolicState.h]:
```suggestion
	smtutil::Expression txConstraints(FunctionDefinition const* _function = nullptr) const;
```

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -152,7 +152,7 @@ void CHC::endVisit(ContractDefinition const& _contract)
 	smtutil::Expression zeroes(true);
 	for (auto var: stateVariablesIncludingInheritedAndPrivate(_contract))
 		zeroes = zeroes && currentValue(*var) == smt::zeroValue(var->type());
-	addRule(smtutil::Expression::implies(initialConstraints(_contract) && zeroes, predicate(entry)), entry.functor().name);
+	addRule(smtutil::Expression::implies((initialConstraints(_contract) && zeroes) && state().txConstraints(nullptr), predicate(entry)), entry.functor().name);
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp]:
```suggestion
	addRule(smtutil::Expression::implies((initialConstraints(_contract) && zeroes) && state().txConstraints(), predicate(entry)), entry.functor().name);
```
after the other suggestion

File: libsolidity/formal/SymbolicState.h (hunk 1 above)
Comment (blishko) [reply libsolidity/formal/SymbolicState.h]: Fixed.

File: libsolidity/formal/CHC.cpp (hunk 2 above)
Comment (blishko) [reply libsolidity/formal/CHC.cpp]: Fixed.

**Comment by leonardoalt** [comment]
@blishko we merged the z3 update, so please install z3 4.8.10 and rebase this, hopefully the crashes will be gone

**Comment by blishko** [comment]
After the upgrade of Z3, there are no more crashes.
Also, it seems that adding constraints for intermediate expressions in checked blocks solved some of the issues mentioned in #10703.
However, the constraints are still missing in unchecked blocks, so I still believe we should add them to the encoding of the deployment phase as proposed in this PR.
I will try to come up with more tests and check BMC.

**Comment by leonardoalt** [comment]
Wait, what's the connection to checked/unchecked blocks?

**Comment by blishko** [comment]
```
void SMTEncoder::defineExpr(Expression const& _e, smtutil::Expression _value)
{
	createExpr(_e);
	solAssert(_value.sort->kind != smtutil::Kind::Function, "Equality operator applied to type that is not fully supported");
	m_context.addAssertion(expr(_e) == _value);

	if (
		auto type = _e.annotation().type;
		m_checked && smt::isNumber(*type)
	)
		m_context.addAssertion(smtutil::Expression::implies(
			currentPathConditions(),
			smt::symbolicUnknownConstraints(expr(_e), type)
		));
}
```
We have this in `defineExpr`, which adds type constraints on Solidity expressions in `checked` blocks, but not in `unchecked` blocks.

**Comment by leonardoalt** [comment]
Ah right, I added that specifically to "automatically" have the type constraints for arithmetic results without having to add anything to the context inside `arithmeticOperator`

File: test/libsolidity/smtCheckerTests/imports/import_base.sol [hunk 3]
```diff
@@ -13,16 +13,10 @@ pragma experimental SMTChecker;
 import "base";
 contract Der is Base {
 	function g(uint y) public {
+		require(x < 10);
```
Comment (blishko) [inline test/libsolidity/smtCheckerTests/imports/import_base.sol]: Added restriction to work around Spacer's nondeterminism

**Comment by leonardoalt** [comment]
Looks good overall, the only thing I'm not sure are the examples with inheritance and `msg.value` being used in different contracts in the hierarchy. Did you try those on Remix as well?

**Comment by blishko** [comment]
I have not added any tests with `msg.value` and inheritance hierarchy, but I should add some.
I will test the examples you suggested in our discussion.

File: test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol [hunk 4]
```diff
@@ -0,0 +1,17 @@
+pragma experimental SMTChecker;
+contract A {
+    uint x = msg.value;
+    constructor() {
+        assert(x == 0); // should hold
+    }
+}
```
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:7]: @blishko you did add some already, like this one for example

Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:7]: There is no inheritance there, `A` and `B` are independent.
That just tests that with a payable constructor, `msg.value` can be anything, but with a non-payable constructor, it must be `0`. In the latter case, I needed to introduce state variable, because the compiler does not even allow to access `msg.value` if the constructor is not payable.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:7]: Ah true

Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:7]: I have added tests with inheritance, `special/msg_value_inheritance_1.sol` and `special/msg_value_inheritance_2.sol`.

File: libsolidity/formal/BMC.cpp [hunk 5]
```diff
@@ -162,7 +162,9 @@ bool BMC::visit(FunctionDefinition const& _function)
 	{
 		reset();
 		initFunction(_function);
-		m_context.addAssertion(m_context.state().txConstraints(_function));
+		m_context.addAssertion(m_context.state().txConstraints(&_function));
+		if (_function.isConstructor() && !_function.isPayable())
+			m_context.addAssertion(m_context.state().txMember("msg.value") == 0);
```
Comment (bshastry) [inline libsolidity/formal/BMC.cpp]: Note to self: Non payable constructor must have zero msg value

File: libsolidity/formal/CHC.cpp [hunk 6]
```diff
@@ -196,8 +196,12 @@ void CHC::endVisit(ContractDefinition const& _contract)
 	connectBlocks(m_currentBlock, summary(_contract));
 
 	setCurrentBlock(*m_constructorSummaries.at(&_contract));
-	m_queryPlaceholders[&_contract].push_back({smtutil::Expression(true), errorFlag().currentValue(), m_currentBlock});
-	connectBlocks(m_currentBlock, interface(), errorFlag().currentValue() == 0);
+	auto constructor = _contract.constructor();
+	auto txConstraints = state().txConstraints();
+	if (!constructor || !constructor->isPayable())
+		txConstraints = txConstraints && state().txMember("msg.value") == 0;
```
Comment (bshastry) [inline libsolidity/formal/CHC.cpp]: If contract contains no constructor or contains a non-payable constructor, msg value must be zero. Why is this constraint added during end visit of contract though?

File: libsolidity/formal/CHC.cpp [hunk 7]
```diff
@@ -262,7 +266,7 @@ void CHC::endVisit(FunctionDefinition const& _function)
 	{
 		auto sum = summary(_function);
 		auto ifacePre = smt::interfacePre(*m_interfaces.at(m_currentContract), *m_currentContract, m_context);
-		auto txConstraints = m_context.state().txConstraints(_function);
+		auto txConstraints = m_context.state().txConstraints(&_function);
```
Comment (bshastry) [inline libsolidity/formal/CHC.cpp]: I guess this is because of API change in txConstraints that now requires a ptr instead of a reference? Could this be a different commit in this PR or is it non-separable from the aim stated in the PR title?

File: libsolidity/formal/SymbolicState.cpp [hunk 8]
```diff
@@ -126,7 +126,7 @@ smtutil::Expression SymbolicState::txMember(string const& _member) const
 	return m_tx.member(_member);
 }
 
-smtutil::Expression SymbolicState::txConstraints(FunctionDefinition const& _function) const
+smtutil::Expression SymbolicState::txConstraints(FunctionDefinition const* _function) const
```
Comment (bshastry) [inline libsolidity/formal/SymbolicState.cpp]: Not sure I understand the motivation for this change.

File: test/libsolidity/smtCheckerTests/abi/abi_encode_packed_simple.sol [hunk 9]
```diff
@@ -23,10 +23,6 @@ contract C {
...
 // Warning 6328: (354-384): CHC: Assertion violation happens here.
-// Warning 1218: (451-481): CHC: Error trying to invoke SMT solver.
-// Warning 6328: (451-481): CHC: Assertion violation might happen here.
+// Warning 6328: (451-481): CHC: Assertion violation happens here.
 // Warning 6328: (560-590): CHC: Assertion violation happens here.
-// Warning 1218: (609-639): CHC: Error trying to invoke SMT solver.
-// Warning 6328: (609-639): CHC: Assertion violation might happen here.
-// Warning 4661: (451-481): BMC: Assertion violation happens here.
-// Warning 4661: (609-639): BMC: Assertion violation happens here.
+// Warning 6328: (609-639): CHC: Assertion violation happens here.
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/abi/abi_encode_packed_simple.sol:28]: Note to self: Change in expectation is possibly unrelated to this PR in that the test case does not contain transaction variables as such.

File: test/libsolidity/smtCheckerTests/abi/abi_encode_packed_string_literal.sol [hunk 10]
```diff
@@ -31,7 +30,6 @@ contract C {
 // Warning 6328: (600-630): CHC: Assertion violation might happen here.
 // Warning 1218: (686-716): CHC: Error trying to invoke SMT solver.
 // Warning 6328: (686-716): CHC: Assertion violation might happen here.
-// Warning 4661: (258-288): BMC: Assertion violation happens here.
 // Warning 4661: (342-372): BMC: Assertion violation happens here.
 // Warning 4661: (515-545): BMC: Assertion violation happens here.
 // Warning 4661: (600-630): BMC: Assertion violation happens here.
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/abi/abi_encode_packed_string_literal.sol:35]: NTS: No tx vars in this test case.

File: test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+contract B {
+    constructor() payable {
+        assert(msg.value == 0); // should fail
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:11]: I may be missing something here but is it illegal to pass zero value while deploying a contract with a payable constructor?

Comment (bshastry) [reply test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:11]: Ah sorry, it fails because `msg.value >= 0` is the correct constraint in this case. Could you please add one such test case for a passing test if one does not exist already?

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol [hunk 12]
```diff
@@ -0,0 +1,57 @@
+pragma experimental SMTChecker;
+
+contract C {
+	constructor() payable {
+		assert(tx.origin >= address(0));
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:5]: May have missed something. Does this PR also add constraints for `tx.origin` or is this test case for a previously added constraint that was missing a test case?

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+		assert(block.gaslimit <= 2**256 - 1);
+		assert(block.number >= 0);
+		assert(block.number <= 2**256 - 1);
+	}
+}
+
+contract D {
+	constructor() payable {
+		unchecked {
+			assert(tx.origin >= address(0));
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:32]: So `unchecked` blocks are essentially disabling SMT checker?

File: test/libsolidity/smtCheckerTests/imports/import_base.sol (hunk 3 above)
Comment (bshastry) [reply test/libsolidity/smtCheckerTests/imports/import_base.sol]: Sounds like a fix for a different issue? Could you please add a comment in the test case describing motivation?

File: test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol [hunk 14]
```diff
@@ -9,5 +9,7 @@ contract C {
 		assert(c > a);
 	}
 }
+// ====
+// SMTIgnoreCex: yes
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol:13]: Out of curiosity: What does the `SMTIgnoreCex` flag do?

File: test/libsolidity/smtCheckerTests/special/msg_sender_3.sol [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+	}
+}
+
+contract D {
+	address owner;
+
+	constructor() {
+		unchecked {
+			owner = msg.sender;
+			assert(owner >= address(0)); // should hold
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/msg_sender_3.sol:18]: Is the comment here valid since it is inside an unchecked block that is presumably unchecked?

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+contract A {
+	uint public x = msg.value;
+	constructor() payable {
+		assert(x == 0); // should fail, A can be constructed with any msg.value
+	}
+}
+
+contract C is A {
+	uint public v = msg.value; // 1
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol]:
```suggestion
	uint public v = msg.value;
```

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+	constructor() A() {
+		assert(v == 0); // should hold
+	}
+}
+// ----
+// Warning 6328: (101-115): CHC: Assertion violation happens here.\nCounterexample:\nx = 1\n\nTransaction trace:\nA.constructor(){ value: 1 }
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol:17]: Why is `CHC` warning in spite of the constraint being added to silence exactly this warning?

Comment (bshastry) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol:17]: Ah I think I understand: The warning accounts for base contract having a payable constructor.

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+contract C is A {
+	uint public v = msg.value; // 1
+	constructor() A() payable {
+		assert(v == 0); // should fail, C can be constructed with any msg.value
+	}
+}
+// ----
+// Warning 6328: (93-107): CHC: Assertion violation happens here.\nCounterexample:\nv = 0, x = 1\n\nTransaction trace:\nC.constructor(){ value: 1 }
+// Warning 6328: (273-287): CHC: Assertion violation happens here.\nCounterexample:\nv = 1, x = 1\n\nTransaction trace:\nC.constructor(){ value: 1 }
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol:18]: Nice! Wdyt about following test configuration?

contract A: non payable
contract B: non payable
contract C -> A, B: payable

I assume `msg.value == 0` cannot hold in any of the constructors, right?

Another test config

contract A: payable
contract B -> A: non payable
contract C -> B, A: non payable

File: test/libsolidity/smtCheckerTests/special/timestamp_2.sol [hunk 19]
```diff
@@ -0,0 +1,9 @@
+pragma experimental SMTChecker;
+
+contract C {
+	uint x;
+
+	constructor() {
+		x = block.timestamp + 0; // No overflow should be reported
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/timestamp_2.sol:7]: Could you add a test for `block.timestamp + 1` if it does not exist already?

**Review by bshastry — COMMENTED** [review]
Perhaps split into a separate PR/commit for newly added tests for constraints that were already being supported and one for spacer fix?

File: test/libsolidity/smtCheckerTests/abi/abi_encode_packed_simple.sol (hunk 9 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/abi/abi_encode_packed_simple.sol:28]: It is true that the test case does not touch transaction variables, but the constraints are nevertheless added in the encoding. So the encoding with this PR is different than the previous encoding. Basically, Spacer sees (syntactically) different problem. And this is known to cause different behaviour in Spacer; sometimes in our favour, sometimes in the wrong direction...

File: test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol (hunk 11 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/msg_value_4.sol:11]: It is part of the `range_check.sol ` test case.

File: test/libsolidity/smtCheckerTests/imports/import_base.sol (hunk 3 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/imports/import_base.sol]: It's a fix for nondeterministic behaviour of Spacer on different platforms on this particular example.
We encounter the problem quite often if there is a change in the encoding (as is the case of this PR).

Basically, with the change in the encoding of this PR, Spacer exhibits different behaviour on different platforms on this example, so it would be impossible to specify a single expected outcome.
So I added the constraint to restrict the search space in the hope of suppressing the non-determinism. 
I can add a comment with the explanation.

File: test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol (hunk 14 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol:13]: It removes the counter-example trace from the output of the tests. It has been added by @leonardoalt to work around the non-determinism in Spacer (which quite often gives different counter-example on different platforms, when the test allows different CEXs).

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol (hunk 17 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_1.sol:17]: Yes, since the base contract has payable constructor, it can be constructed using any `msg.value`, so the assertion is violated when the contract is constructed with value greater than 0.

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol (hunk 18 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol:18]: Yes, the situation should be the same as in the two-contract inheritance covered by `msg_value_inheritance_1.sol` and `msg_value_inheritance_2.sol`. Not sure if the tests would bring additional value, but I can add them if you believe so.

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol (hunk 13 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:32]: It is not disabled, but there is a difference in encoding, see, e.g., https://github.com/ethereum/solidity/pull/10704#issuecomment-769304388

Basically, in checked blocks type constraints are automatically added for intermediate expressions, which as a side effect constrains also transaction variables. This change has been made after the first version of this PR.

File: test/libsolidity/smtCheckerTests/special/msg_sender_3.sol (hunk 15 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_sender_3.sol:18]: `assert` is still checked and the type constraints are valid also in unchecked blocks.

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol (hunk 12 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:5]: Yes, this PR adds also type constraints for `tx.origin` *in deployment phase*, which were missing before.

File: test/libsolidity/smtCheckerTests/imports/import_base.sol (hunk 3 above)
Comment (bshastry) [reply test/libsolidity/smtCheckerTests/imports/import_base.sol]: Sounds good! It would be nice to document this somehow as a comment imo.

File: libsolidity/formal/SymbolicState.cpp (hunk 8 above)
Comment (blishko) [reply libsolidity/formal/SymbolicState.cpp]: The idea is that I need to obtain `txConstraints` also for the deployment phase, and also for the situation where no explicit constructor is present, because the transaction data might be used in implicit constructor using direct initiatilization of the state variables.
In that case,  there is no `FunctionDefinition` to pass to this method, so I wanted to allow passing `nullptr` for this case.

Comment (bshastry) [reply libsolidity/formal/SymbolicState.cpp]: Okay, was thinking if an assert on `_function` is needed but I just saw `_function && ..` in the if clause so that should take care of it. I didn't find any potential for nullptr deref.

File: libsolidity/formal/CHC.cpp (hunk 7 above)
Comment (blishko) [reply libsolidity/formal/CHC.cpp]: Currently, it is needed to allow the use of function `txConstraints` also for the deployment phase where there is no explicit constructor.
But now I am thinking that we could split `txConstraints` to two methods, one that does require `FunctionDefinition` (adding constraints on function signature) and one that does not require it (only adding the type constrains on blockchain and transaction variables independent of the function being called). @leonardoalt, what do you think?

File: test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol (hunk 14 above)
Comment (bshastry) [reply test/libsolidity/smtCheckerTests/operators/conditional_assignment_3.sol:13]: Ah I see. Cex presumably stands for Counter*ex*ample :)

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol (hunk 12 above)
Comment (bshastry) [reply test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:5]: okay, I looked at https://github.com/ethereum/solidity/pull/10704/files#diff-fa0f6553d9da37a75e07770a2233fabd4c32b7b91b96929b5f42c7fab0ec696cR167 and the positive diff only covered placing a constraint on `msg.value` so wasn't sure. Could you please link the lines where constraints on `tx.origin` is added? Thanks!

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol (hunk 18 above)
Comment (bshastry) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol:18]: I'm not sure if it will bring value to be honest, was just thinking out loud :)

```
contract A {
  uint x = msg.value;
  constructor() {
    assert(x == 0);
  }
}
contract B {
  constructor() payable {
    assert(msg.value >= 0);
  }
}
contract C is A, B {
  constructor() A() B() payable {
    assert(msg.value >= 0);
  }
}
```

I was wondering which of the assertions above ~~are thrown~~ result in warnings

File: test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol (hunk 12 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/bmc_coverage/range_check.sol:5]: You are right than in BMC only the constraint on `msg.value` has been added, but all constraints were added for CHC:
https://github.com/ethereum/solidity/pull/10704/files#diff-42bd29aaf4d31608e91e7940d97d1710c3c6bf6f860bf70f5885afe8b96f68b9

File: libsolidity/formal/CHC.cpp [hunk 20]
```diff
@@ -196,8 +196,12 @@ void CHC::endVisit(ContractDefinition const& _contract)
 	connectBlocks(m_currentBlock, summary(_contract));
 
 	setCurrentBlock(*m_constructorSummaries.at(&_contract));
-	m_queryPlaceholders[&_contract].push_back({smtutil::Expression(true), errorFlag().currentValue(), m_currentBlock});
-	connectBlocks(m_currentBlock, interface(), errorFlag().currentValue() == 0);
+	auto constructor = _contract.constructor();
+	auto txConstraints = state().txConstraints();
```
Comment (bshastry) [inline libsolidity/formal/CHC.cpp]: Where are these constraints added?

File: libsolidity/formal/CHC.cpp (hunk 7 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Yea I think splitting could be a good idea. But the current version is also fine, I don't have a strong preference

File: libsolidity/formal/CHC.cpp (hunk 20 above)
Comment (blishko) [reply libsolidity/formal/CHC.cpp]: Just below (lines 203 and 204), as part of the condition for the placeholder and for the edge connecting blocks.

Comment (bshastry) [reply libsolidity/formal/CHC.cpp]: Ah I see, sorry I misunderstood. So `state().txConstraints()` existed but was only added in this PR. I was looking for the origin of `state().txConstraints()` :)

Comment (blishko) [reply libsolidity/formal/CHC.cpp]: Yes, the code existed before, and the constraints have already been added for regular transactions before, but they were missing from the deployment phase, to capture the fact that they must hold also when the transaction data are accessed in constructor.

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol (hunk 18 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_2.sol:18]: Only the first one is reported, it can be violated when `A` is constructed  as part of `C`. I am adding this test case.

File: test/libsolidity/smtCheckerTests/special/timestamp_2.sol (hunk 19 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/timestamp_2.sol:7]: Added.

File: test/libsolidity/smtCheckerTests/imports/import_base.sol (hunk 3 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/imports/import_base.sol]: I've added the comment in the test case.

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+contract B {
+	constructor() payable {
+		assert(msg.value >= 0); // should hold
+	}
+}
+contract C is A, B {
+	constructor() A() B() payable {
+		assert(msg.value >= 0); // should hold
+	}
+}
```
Comment (bshastry) [inline test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol:19]: Just curious what happens with the following

```
pragma experimental SMTChecker;

contract A {
	uint public x = msg.value;
	constructor() {
		assert(x == 0); // can fail when A is constructed as part of C or B
	}
}

contract B is A {
	constructor() A() payable {
		assert(msg.value >= 0); // should hold
	}
}
contract C is B, A {
	constructor() B() A() payable {
		assert(msg.value >= 0); // should hold
	}
}
```

and 

```
pragma experimental SMTChecker;

contract A {
	constructor() payable {
		assert(x == 0); // should hold when constructed via C, fail otherwise?
	}
}

contract B is A {
	constructor() A() payable {
		assert(msg.value >= 0); // should hold
	}
}
contract C is A, B {
	constructor() A() B() {
		assert(msg.value == 0); // should hold
	}
}
```

Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol:19]: These examples do not compile, the inheritance graph is not linearizable. And in the second example, definition of `x` is missing.

Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol:19]: @bshastry , @leonardoalt , I have split the function `txConstraints` to `txTypeConstraints` and `txFunctionConsttraints`, I think it makes sense.
This also showed that previously the `non-payable` constraint `msg.value == 0` was missing for `fallback` function. 
It also forms the conjunction of the constraints in a slightly different way, so there are a bunch of changes in command line tests where the BMC formula is printed to the output.

@bshastry , I am not sure if I have addressed all your comments, let me know I missed something.

Comment (bshastry) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol:19]: @blishko ah I should've tried those test cases before typing them out. Here you go

```
pragma experimental SMTChecker;

contract A {
	uint public x = msg.value;
	constructor() {
		assert(x == 0); // can fail when A is constructed as part of C or B
	}
}

contract B is A {
	constructor() A() payable {
		assert(msg.value >= 0); // should hold
	}
}
contract C is A, B {
	constructor() B() payable {
		assert(msg.value >= 0); // should hold
	}
}
```

```
pragma experimental SMTChecker;

contract A {
        uint x;
	constructor() payable {
		assert(x == 0); // should hold when constructed via C, fail otherwise?
	}
}

contract B is A {
	constructor() A() payable {
		assert(msg.value >= 0); // should hold
	}
}
contract C is A, B {
	uint y = msg.value;
	constructor() B() {
		assert(y == 0); // should hold
	}
}
```

**Comment by bshastry** [comment]
LGTM. I feel the test commits may be squashed.

File: Changelog.md [hunk 22]
```diff
@@ -7,7 +7,7 @@ Compiler Features:
 
 
 Bugfixes:
-
+ * SMTChecker: Fix missing type constraints on blockchain variables in the deployment phase.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * SMTChecker: Fix missing type constraints on block and transaction variables in the deployment phase.
```

File: libsolidity/formal/SymbolicState.h [hunk 23]
```diff
@@ -126,7 +126,9 @@ class SymbolicState
 	smtutil::SortPointer const& txSort() const { return m_tx.sort(); }
 	void newTx() { m_tx.newVar(); }
 	smtutil::Expression txMember(std::string const& _member) const;
-	smtutil::Expression txConstraints(FunctionDefinition const& _function) const;
+	smtutil::Expression txFunctionConstraints(FunctionDefinition const& _function) const;
+	smtutil::Expression txTypeConstraints() const;
+	smtutil::Expression txNonPaybleConstraint() const;
```
Comment (leonardoalt) [inline libsolidity/formal/SymbolicState.h]:
```suggestion
	smtutil::Expression txNonPayableConstraint() const;
```

File: test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol (hunk 21 above)
Comment (blishko) [reply test/libsolidity/smtCheckerTests/special/msg_value_inheritance_3.sol:19]: Regarding the first example, it is as you write in the comments. The assertion in `A` can fail when constructed as partt of `C` or `B`. The other assertions hold. I think this is the same situation as in the test case `msg_value_inheritance_2.sol`.

Regarding the second example, as you have written it, all assertions hold. If you meant `uint x = msg.value` then the assertion in `A` can fail when contract `A` is constructed with value greater than 0, which is possible since constructor of `A` is `payable`.

File: libsolidity/formal/SymbolicState.h (hunk 23 above)
Comment (blishko) [reply libsolidity/formal/SymbolicState.h]: Fixed.

File: Changelog.md (hunk 22 above)
Comment (blishko) [reply Changelog.md]: Changed.

**Comment by blishko** [comment]
Rebased and squashed.

**Review by leonardoalt — APPROVED** [review]
lgtm, but I'll wait for @bshastry

**Review by bshastry — APPROVED** [review]
LGTM too
